test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewFeatures/ViewDataDictionaryOfTModelTest.cs

```diff
@@ -223,6 +226,48 @@ public void CopyConstructors_ThrowInvalidOperation_IfModelIncompatible()
             Assert.Equal(expectedMessage, exception.Message);
         }
 
+        [Theory]
+        [InlineData(null)]
+        [InlineData(23)]
+        public void CopyConstructor_DoesNotChangeMetadata_WhenValueCompatibleWithSourceMetadata(int? model)
+        {
+            // Arrange
+            var metadataProvider = new EmptyModelMetadataProvider();
+            var source = new ViewDataDictionary<int?>(metadataProvider)
+            {
+                Model = -48,
+            };
+
+            // Act
+            var viewData = new ViewDataDictionary<int?>(source, model);
+
+            // Assert
+            Assert.NotNull(viewData.ModelExplorer);
+            Assert.NotSame(source.ModelExplorer, viewData.ModelExplorer);
+            Assert.Same(source.ModelMetadata, viewData.ModelMetadata);
```

Comment:
This assertion fails w/o the `src` changes when `model==23`. `viewData.ModelMetadata.ModelType` is `typeof(int)`.